### PR TITLE
[CLD-7046] Ability to disable self hosted server-side requests to CWS 

### PR DESCRIPTION
**Pull request by nickmisasi**
#### Summary
In air-gapped environments, server side requests made to the CWS will always fail. Currently, these requests will result in 500's and errors in the mattermost error log. This PR adds the ability to disable requests to the CWS entirely, instead returning a 400 with a message about the system administrator disabling the functionality.

#### Ticket Link
https://mattermost.atlassian.net/browse/CLD-7046

#### Release Note

```release-note
New Config CloudSettings.Disable (via config.json, or environment variable), default false. When set to true, all requests to the Mattermost Customer Portal from a workspace will be disabled
```

### CODE REVIEW

**Review by vish9812 — APPROVED** [review]

**Comment by nickmisasi** [comment]
@hanzei gentle ping on this one - if you'd rather me pass the review to someone else please let me know!

File: server/channels/api4/cloud.go [hunk 1]
```diff
@@ -65,10 +65,15 @@ func (api *API) InitCloud() {
 
 func ensureCloudInterface(c *Context, where string) bool {
 	cloud := c.App.Cloud()
+	disabled := c.App.Config().CloudSettings.Disable != nil && *c.App.Config().CloudSettings.Disable
```
Comment (gabrieljackson) [inline server/channels/api4/cloud.go:68]: nit: could move this after the `cloud == nil` check, but not a big deal.

File: server/channels/api4/cloud.go [hunk 2]
```diff
@@ -65,10 +65,15 @@ func (api *API) InitCloud() {
...
 	if cloud == nil {
 		c.Err = model.NewAppError(where, "api.server.cws.needs_enterprise_edition", nil, "", http.StatusBadRequest)
 		return false
 	}
+	if disabled {
+		c.Err = model.NewAppError(where, "api.server.cws.disabled", nil, "", http.StatusBadRequest)
```
Comment (gabrieljackson) [inline server/channels/api4/cloud.go:74]: I am 0/5 on changing this, but is `http.StatusBadRequest` what we want to return? Just curious if we had a reason for this over other returns.

**Review by gabrieljackson — APPROVED** [review]
Looks good. Just one non-blocking nit and a question.

File: server/channels/api4/cloud.go (hunk 1 above)
Comment (hanzei) [reply server/channels/api4/cloud.go:68]: :+1: for checking `disabled` first

File: server/channels/api4/cloud.go (hunk 2 above)
Comment (hanzei) [reply server/channels/api4/cloud.go:74]: `BadRequest` might not be the right one. Per https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400 it's a

> response status code indicates that the server cannot or will not process the request due to something that is perceived to be a client error (for example, malformed request syntax, invalid request message framing, or deceptive request routing).


Maybe 422 Unprocessable Content is better here as it is about semantic errors and 400 about syntactic ones.

> The HyperText Transfer Protocol (HTTP) 422 Unprocessable Content response status code indicates that the server understands the content type of the request entity, and the syntax of the request entity is correct, but it was unable to process the contained instructions.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422

**Review by hanzei — APPROVED** [review]
Nice work. +1: Sorry for the delay; the PR got lost in my review queue.

Let"s ensure to properly comment on what admins need to configure in air-gapped environments.

cc @cwarnermm

**Comment by cwarnermm** [comment]
@nickmisasi - As per the release note "New Config CloudSettings.Disable, default false. When set to true, all requests to the Mattermost Customer Portal from a workspace will be disabled" -- where would an admin be setting this config? System Console?

**Comment by nickmisasi** [comment]
@cwarnermm config.json or environment variable will be the only way to set this

**Comment by cwarnermm** [comment]
@nickmisasi - Would it be safe to assume that Cloud admins won't be performing this change via ``config.json``/env var?

**Comment by nickmisasi** [comment]
@cwarnermm correct, cloud admins won't be changing this config. You just triggered a thought and I've just pushed a commit - so now, not only will they not be changing, it won't be possible for them to do so.

**Comment by cwarnermm** [comment]
@nickmisasi - Is this new config Cloud-specific, or would self-hosted admins potentially set this as well? If it's the latter, it's something we'll add to docs.

**Comment by nickmisasi** [comment]
@cwarnermm this config is exclusive to self-hosted environments. It's intention is for those that are air gapped

**Comment by cwarnermm** [comment]
Glad I asked, @nickmisasi - I had interpreted it backwards :) Thanks!

**Pull request merged**